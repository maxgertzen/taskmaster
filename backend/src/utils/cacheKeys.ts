export const CACHE_KEYS = {
  TASK: (userId: string, taskId: string) =>
    `cache:user:${userId}:task:${taskId}` as const,
  TASK_LIST: (userId: string, listId: string) =>
    `cache:user:${userId}:tasks:${listId}` as const,
  LIST: (userId: string, listId: string) =>
    `cache:user:${userId}:list:${listId}` as const,
  LISTS: (userId: string) => `cache:user:${userId}:lists` as const,
  SEARCH_RESULTS: (userId: string, query: string) =>
    `cache:user:${userId}:search:${query}` as const,
} as const;
