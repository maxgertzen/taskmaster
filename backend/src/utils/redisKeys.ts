export const REDIS_KEYS = {
  USER: (userId: string) => `user:${userId}` as const,
  TASK: (userId: string, taskId: string) =>
    `user:${userId}:task:${taskId}` as const,
  TASK_LIST: (userId: string, listId: string) =>
    `user:${userId}:tasks:${listId}` as const,
  TASKS_INDEX: "tasks_index" as const,
  LIST: (userId: string, listId: string) =>
    `user:${userId}:list:${listId}` as const,
  LISTS: (userId: string) => `user:${userId}:lists` as const,
} as const;
