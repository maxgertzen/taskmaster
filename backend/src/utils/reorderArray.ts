export const reorderArray = <T>(
  array: T[],
  fromIndex: number,
  toIndex: number
): T[] => {
  const arr = [...array];
  const [movedItem] = arr.splice(fromIndex, 1);
  arr.splice(toIndex, 0, movedItem);
  return arr;
};
