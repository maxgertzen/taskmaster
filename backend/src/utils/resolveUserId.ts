import { getAppContainer } from "@src/container";

export async function resolveUserId(
  auth0Id: string,
  email?: string,
  name?: string
): Promise<string> {
  const container = getAppContainer();
  const usersService = container.cradle.usersService;
  const user = await usersService.getOrCreateUser(auth0Id, email, name);

  if (!user) {
    throw new Error(`User with auth0Id ${auth0Id} not found`);
  }

  if ("_id" in user) {
    return user._id?.toString() ?? "";
  }

  return user.auth0Id;
}
