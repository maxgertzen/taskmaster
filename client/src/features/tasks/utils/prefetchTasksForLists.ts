import { QueryClient } from '@tanstack/react-query';

import { QUERY_KEYS } from '@/shared/api/query-keys';
import { STALE_TIME } from '@/shared/constants/staleTime';

import { fetchTasks } from '../api';

export const prefetchTasksForLists = (
  listIds: string[],
  queryClient: QueryClient,
  token: string | null
) => {
  if (!token || !listIds.length) return;

  listIds.forEach((listId) => {
    queryClient.prefetchQuery({
      queryKey: QUERY_KEYS.tasks({ listId }),
      queryFn: fetchTasks(token),
      staleTime: STALE_TIME,
    });
  });
};
