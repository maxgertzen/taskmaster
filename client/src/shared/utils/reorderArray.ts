export const reorderArray = <T>(
  array: T[],
  fromIndex: number,
  toIndex: number
): T[] => {
  const reordered = [...array];
  const [movedItem] = reordered.splice(fromIndex, 1);
  reordered.splice(toIndex, 0, movedItem);
  return reordered;
};
