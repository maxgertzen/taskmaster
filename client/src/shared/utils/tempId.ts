export const generateTempId = (prefix: string = 'temp') =>
  `${prefix}-${Date.now()}`;
