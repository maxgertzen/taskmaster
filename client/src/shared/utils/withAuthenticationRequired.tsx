import { withAuthenticationRequired as realWithAuthenticationRequired } from '@auth0/auth0-react';

import { IS_AUTH0_DISABLED } from '@/constants';
import { mockWithAuthenticationRequired } from '@/mocks';

const isMock = IS_AUTH0_DISABLED === 'true';

export const withAuthenticationRequired = isMock
  ? mockWithAuthenticationRequired
  : realWithAuthenticationRequired;
