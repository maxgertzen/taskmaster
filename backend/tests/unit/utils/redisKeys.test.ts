import { REDIS_KEYS } from "@src/utils/redisKeys";
import { testConfig } from "@tests/data/utils";

const { validInputs, edgeCases } = testConfig;

describe("REDIS_KEYS", () => {
  describe("USER key", () => {
    it("should generate correct user key", () => {
      const result = REDIS_KEYS.USER(validInputs.userId);
      expect(result).toBe(`user:${validInputs.userId}`);
    });

    it("should handle empty userId", () => {
      const result = REDIS_KEYS.USER(edgeCases.emptyStrings.userId);
      expect(result).toBe("user:");
    });
  });

  describe("TASK key", () => {
    it("should generate correct task key", () => {
      const result = REDIS_KEYS.TASK(validInputs.userId, validInputs.taskId);
      expect(result).toBe(
        `user:${validInputs.userId}:task:${validInputs.taskId}`
      );
    });

    it("should handle special characters", () => {
      const { userId, taskId } = edgeCases.specialChars;
      const result = REDIS_KEYS.TASK(userId, taskId);
      expect(result).toBe(`user:${userId}:task:${taskId}`);
    });
  });

  describe("TASK_LIST key", () => {
    it("should generate correct task list key", () => {
      const result = REDIS_KEYS.TASK_LIST(
        validInputs.userId,
        validInputs.listId
      );
      expect(result).toBe(
        `user:${validInputs.userId}:tasks:${validInputs.listId}`
      );
    });
  });

  describe("LIST key", () => {
    it("should generate correct list key", () => {
      const result = REDIS_KEYS.LIST(validInputs.userId, validInputs.listId);
      expect(result).toBe(
        `user:${validInputs.userId}:list:${validInputs.listId}`
      );
    });
  });

  describe("LISTS key", () => {
    it("should generate correct lists key", () => {
      const result = REDIS_KEYS.LISTS(validInputs.userId);
      expect(result).toBe(`user:${validInputs.userId}:lists`);
    });
  });

  describe("TASKS_INDEX", () => {
    it("should return correct tasks index key", () => {
      expect(REDIS_KEYS.TASKS_INDEX).toBe("tasks_index");
    });
  });

  it("should match snapshot", () => {
    expect(REDIS_KEYS).toMatchSnapshot();
  });
});
