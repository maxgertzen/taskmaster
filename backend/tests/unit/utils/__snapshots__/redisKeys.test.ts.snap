// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`REDIS_KEYS should match snapshot 1`] = `
{
  "LIST": [Function],
  "LISTS": [Function],
  "TASK": [Function],
  "TASKS_INDEX": "tasks_index",
  "TASK_LIST": [Function],
  "USER": [Function],
}
`;
